Serving iCal data for client to fetch on page load

EthanThatOneKid opened the pull request:
This will allow to render the event data every time the user enters the page.

EthanThatOneKid committed:
Copy event URL implemented 🔗

Created algorithm to make unique event fragments to be copied from the title of each event item. Each event gets its own, unique fragment that can be copied and linked to.

bjwarfield committed:
Add Carousel Button Functionality

Hide standard scrollbar on carousel.
Added custom Horizontal Scroll Button to carousel.
Add linear gradient effect to edge of carousel.
Hide Scroll buttons on mobile view.

EthanThatOneKid committed:
Merge pull request #102 from bjwarfield/launch/events-route

Add Carousel Button Functionality

EthanThatOneKid committed:
Update event-carousel.svelte

Inspired by @bjwarfield's changes, I took it upon myself to refactor the code. There is always more room for improvement, so future changes are definitely encouraged.

EthanThatOneKid committed:
Updated event link generator

Resolves comment <https://github.com/EthanThatOneKid/acmcsuf.com/commit/910f4c172c43b700757534aed37edbed33087866#r50030360>.

diamondburned committed:
Highlight selected event

EthanThatOneKid committed:
Final touches to the /events route 🎉

We fixed some bugs before launching the events route.

- Offset for the link when an event's fragment is included in the URL.
- Fixed the "margins" for the event carousel.
- Added a popover for copying the event URL.
- Broke: Grabbing icon when grabbing the event carousel (it works some of the time).

EthanThatOneKid committed:
Update event-item.svelte

EthanThatOneKid committed:
Updated the border and box shadow color to acm-red.

EthanThatOneKid committed:
Update event-item.svelte

Updated the color of the outline and box shadow of the highlighted event.

EthanThatOneKid committed:
Merge branch 'launch/events-route-target' into launch/events-route

EthanThatOneKid committed:
Launching the /events route

Now, the navbar has a link to the /events route, so it will be included in the production site!

EthanThatOneKid committed:
Attempting to lazy load the ICAL data instead of render it server side.

EthanThatOneKid committed:
Update events.svelte

If an error occurs, just say that there are no events.

EthanThatOneKid committed:
Added public API route for event data.

diamondburned commented on src/lib/parse-ical-data.ts:
```diff
@@ -18,25 +71,41 @@ export interface AcmEvent {
 
+const convertIcalDateTime = (datetime: string): Date => {
+  const characters = datetime.split("");
+  const fullYear = characters.splice(0, 4).join("");
```
`characters.splice(a, b).join("")` could be `datetime.slice(a, b)` instead without splitting `datetime`.

EthanThatOneKid replied:
That would possibly be more readable. It would indeed work since the numbers in the iCal datetime format are always in the same spots.

EthanThatOneKid replied:
Fixed: <https://github.com/EthanThatOneKid/acmcsuf.com/pull/105/files#diff-c9b7c8232c884e8d6c62a724e8cf309ba0d316ca3bf4cad5a1515820a0911edeR76-R81>

diamondburned commented on src/routes/events.json.ts:
```diff
@@ -0,0 +1,22 @@
+export async function get(req: Request, res: Response) {
+    const now = Date.now();
+    if (now > eventExpirationTimestamp) {
+        const data = await fetch(ICAL_TARGET_URL).then(response => response.text());
```
Is this still done by the client? Does it need `no-cors` then?

EthanThatOneKid replied:
This is being ran on the server, so we can omit the `no-cors` mode.

diamondburned commented on src/routes/events.svelte:
```diff
@@ -6,35 +6,20 @@
     // Lazily load the ICAL data for the event carousel.
-    fetch(ICAL_TARGET_URL, { mode: "no-cors" })
-      .then((response) => response.text())
+    fetch("../events.json")
```
Where is this pointed to?

EthanThatOneKid replied:
I created a public API endpoint (to act as a proxy server) to see if that _could_ solve the problem. If you go to `acmcsuf.com/events.json`, it will load the array of events.

diamondburned replied:
I'm getting a 404.

Would this require a server then?

EthanThatOneKid replied:
Here is what it looks like based on the preview: <https://acm-csuf-site-cksb0sp2f-ethanthatonekid.vercel.app/events.json>

EthanThatOneKid committed:
Merge branch 'launch/events-route' into fix/events-route-lazy-load

EthanThatOneKid committed:
Added node-fetch to the dependencies.

EthanThatOneKid committed:
Update events.svelte

Forget to uncomment something.

EthanThatOneKid committed:
Added configurations to run the server on Vercel.

EthanThatOneKid committed:
Added an isLoading message

Added an isLoading feature so that the user is not confused by the "There are currently no events scheduled." message while they wait for the data to load. I wanted to use this [await block](https://svelte.dev/docs#await), but it felt right to just leave the loading in the `onMount` callback.
